File: src/main/java/cpsc224/creatures/Player.java
package cpsc224.creatures;

import cpsc224.items.Inventory;
import cpsc224.levels.Coordinate;

import javax.swing.*;

/**
 * A class to represent a player in the game.
 */
public class Player extends Creature {

    private int gold;

    /**
     * Creates a player with the specified name, max health, base modifiers, and inventory
     * @param creatureName the name
     * @param maxHealth the max health
     * @param modifiers the base modifiers
     * @param inventory the inventory
     */
    public Player(String creatureName, int maxHealth, CreatureModifiers modifiers, Inventory inventory) {
        this(creatureName, maxHealth, modifiers, inventory, null);
    }

    /**
     * Creates a player with the specified name, max health, base modifiers, and inventory
     * @param creatureName the name
     * @param maxHealth the max health
     * @param modifiers the base modifiers
     * @param inventory the inventory
     */
    public Player(String creatureName, int maxHealth, CreatureModifiers modifiers, Inventory inventory, ImageIcon icon) {
        super(creatureName, maxHealth, modifiers, inventory, icon);

        gold = 0;
        setCurrentPosition( new Coordinate(0, 0));
    }

    /**
     * Gets the amount of gold.
     * @return amount of gold
     */
    public int getGold() {
        return gold;
    }

    /**
     * Adds the amount of gold.
     * @param amount the amount of gold
     */
    public void addGold(int amount) {
        gold += amount;
    }

    /**
     * Subtracts the amount of gold.
     * @param amount the amount of gold
     */
    public void subractGold(int amount) {
        gold -= amount;
    }

}

File: web/src/game/creatures/Player.ts
import { Creature } from './Creature';
import { CreatureModifiers } from './CreatureModifiers';
import { Inventory } from '../items/Inventory';
import type { IconRef } from '../utils/icons';
import { Coordinate } from '../levels/Coordinate';
import type { PlayerClass } from './playerClass';

export class Player extends Creature {
  private gold = 0;
  private keys = 0;
  private playerClass: PlayerClass | null = null;
  /** Knight passive: consecutive landed Slice hits. */
  private consecutiveSlashHits = 0;
  /** Knight passive: whether Bleed already procced on this slash streak. */
  private slashBleedProcced = false;

  constructor(
    name: string,
    maxHealth: number,
    modifiers: CreatureModifiers,
    inventory: Inventory,
    icon: IconRef = null,
  ) {
    super(name, maxHealth, modifiers, inventory, icon, []);
    this.currentPosition = new Coordinate(0, 0);
  }

  getGold(): number {
    return this.gold;
  }

  addGold(n: number): void {
    this.gold += n;
  }

  getKeys(): number {
    return this.keys;
  }

  addKeys(n: number): void {
    this.keys += n;
  }

  /** Returns false if not enough keys. */
  spendKey(): boolean {
    if (this.keys <= 0) return false;
    this.keys -= 1;
    return true;
  }

  /** Typo preserved from Java Player.subractGold */
  subractGold(n: number): void {
    this.gold -= n;
  }

  getPlayerClass(): PlayerClass | null {
    return this.playerClass;
  }

  setPlayerClass(cls: PlayerClass): void {
    this.playerClass = cls;
  }

  getConsecutiveSlashHits(): number {
    return this.consecutiveSlashHits;
  }

  setConsecutiveSlashHits(n: number): void {
    this.consecutiveSlashHits = Math.max(0, n);
  }

  hasSlashBleedProcced(): boolean {
    return this.slashBleedProcced;
  }

  setSlashBleedProcced(v: boolean): void {
    this.slashBleedProcced = v;
  }

  resetConsecutiveSlashHits(): void {
    this.consecutiveSlashHits = 0;
    this.slashBleedProcced = false;
  }
}
